### ISSUE TITLE
Q: Windows support ...? MSSQL ..?

### ISSUE THREAD

**Issue by ewi2345** [blue-yonder/turbodbc#16]
Is there anything in the code that would prevent turbodbc run on a windows system ?? Is MSSQL already supported ..?

**Comment by MathMagique** [comment]
To be completely honest: I do not know. I have little experience with developing Windows software. There is some hope that it would work (though probably not too much), because the unixODBC header files (ODBC driver on Linux) should be _virtually_ identical to the Windows ODBC header files. Boost is available for Windows as well (you may need to install it from somewhere to a place where Python's build system will find it), so it just _could_ work.

That being said, it is totally untested and I may have accidentally used Linux-specifics somewhere in the code.

Similarly, MSSQL support has not been tested yet. There is a good chance that it would work, given that turbodbc is successfully tested with MySQL, PostgreSQL, and Exasol. The chance is not 100 percent, because some drivers behave slightly differently concerning data types, supported ODBC features, and retrievable meta-information.

So in short: you can give it a shot, but there are no guarantees. I would appreciate your feedback in that matter. Maybe I can invest some time in adding proper Windows support in the upcoming months.

**Comment by keitherskine** [comment]
Currently, when I try to `pip install turbodbc` on Windows 10, I get the unfortunately familiar "error: Microsoft Visual C++ 10.0 is required (Unable to find vcvarsall.bat).", because it's trying to compile the package from source.  To make this work, I would have to download and set up Visual C++ on my PC, which is not as straightforward as it sounds and may require me to buy one or more Visual C++ licenses.  Hence it would be very handy if pre-compiled python wheels could be uploaded to PyPI, which install directly on Windows.  If that is not possible, you could try to persuade Christoph Gohlke to include turbodbc in his excellent set of wheels for Windows, publicly available from http://www.lfd.uci.edu/~gohlke/pythonlibs/.  That repository is rebuilt every day, as far as I can tell, and already includes the packages [pyodbc](http://www.lfd.uci.edu/~gohlke/pythonlibs/#pyodbc) and [ceODBC](http://www.lfd.uci.edu/~gohlke/pythonlibs/#ceodbc).

**Comment by MathMagique** [comment]
As far as I know, Visual Studio Express is free to download and use, and should be enough to compile things. But there may be other things that are missing (such as boost headers, for example). But I have a Windows computer here, so things may not look too bleak.

**Comment by keitherskine** [comment]
Just FYI, different versions of Visual Studio are needed to compile different versions of Python modules.  See here for some background: https://blogs.msdn.microsoft.com/pythonengineering/2016/04/11/unable-to-find-vcvarsall-bat/.  Hence, it's not trivial to pull together all the python wheels necessary for all Python versions, but if you can do that then that would be great.

**Comment by keitherskine** [comment]
I contacted Christoph Gohlke.  He couldn't get turbodbc to compile on Windows.  Here is, in part, his reply as to why not:
```
"Version 1.0.1 does not compile with Visual Studio 2015 for
Python >= 3.5 on Windows. It's missing 'includes' for the Windows
header, it's not 64-bit proof due to use of long types, it makes use
of C++ features not supported by MS compilers, and it tries to
compile a shared library as a Python extension module. I have not
tried the git repro, which seem to have switched to a CMake build."
```

**Comment by MathMagique** [comment]
Ok, so things are looking bleak right now ;-). Obviously, some additional work is required (plus getting myself up to speed with windows development, apparently).

Thanks @keitherskine for contacting Christoph and providing the feedback. And, of course, thanks to Christoph for taking the time to establish the status quo.

**Comment by TWAC** [comment]
Got it compiling and doing basic selects from drivers "PostgreSQL ANSI(x64)" and "SQL Server".
Could not get the cmake machinery working, so I created a simple script for creating the sdist.
Not close to merge quality, but a start.
https://github.com/TWAC/turbodbc/tree/winhack

**Comment by MathMagique** [comment]
@TWAC Great to hear that! From a quick glance, I see that your branch uses `__int64` types. Is it possible to switch to the standard [fixed-width integer types](http://en.cppreference.com/w/cpp/types/integer)?

**Comment by TWAC** [comment]
Yes, int64_t works the same.

**Comment by TWAC** [comment]
I rebased on the unicode branch.
Replaced most long with intptr_t (seems to make sense since ODBC likes to mix pointers and integers).
Replaced most unsigned short with wchar_t (VC++ did not seem to think unsigned short is compatible with SQLWCHAR).

Now I can do bulk select and insert with "PostgreSQL ANSI(x64)", if I first `set CLIENT_ENCODING='UTF8'`.
For "SQL Server" driver I can do bulk selects, but bulk inserts dies with
```turbodbc.exceptions.DatabaseError: ODBC error
state: HY104
native error code: 0
message: [Microsoft][ODBC SQL Server Driver]Invalid scale value
```

**Comment by MathMagique** [comment]
`unsigned short` to `wchar_t` will not work on Linux and OSX, because on these platforms `wchar_t` uses 4 bytes. I'm not sure about the impact the `intptr_t` change. May I suggest that you open a pull request (even if you feel that not all issues are resolved yet). This will give you access to turbodbc's continuous integration build, so you can easily check whether your changes work with Linux / OSX as well.

We need to have a continuous build in place for Windows as well, otherwise any change someone does can break Windows compatibility without noticing. I would pick [AppVeyor](https://www.appveyor.com) for this task.

Concerning the error you see, that's hard to debug without having a windows build or any information considering the data types you used in the destination table and which value you tried to insert.

**Comment by TWAC** [comment]
Added an explicit cast to keep the unsigned shorts.
Splitted my commit into smaller chunks, and opened #57 for buildbots check. (It does build and run for me on Windows and Linux.)

Will see if I can create a repeatable case for the error.

**Comment by TWAC** [comment]
The problem I have is related to datetime parameters, minimal case below.
The same code works in Linux, with the winhack branch and driver "ODBC Driver 13 for SQL Server".

```python
import datetime
from turbodbc import connect
conn = connect(DRIVER="SQL Server", ...)
curs = conn.cursor()
curs.execute("select ?", [datetime.datetime.now()])
```

```
Traceback (most recent call last):
  File "C:\python36\lib\site-packages\turbodbc\exceptions.py", line 31, in wrapper
    return f(*args, **kwds)
  File "C:\python36\lib\site-packages\turbodbc\cursor.py", line 70, in execute
    buffer.add_set(parameters)
turbodbc_intern.Error: ODBC error
state: HY104
native error code: 0
message: [Microsoft][ODBC SQL Server Driver]Invalid scale value
```

**Comment by MathMagique** [comment]
Could you try what happens if you set this value to another one:
https://github.com/blue-yonder/turbodbc/blob/master/cpp/turbodbc/Library/src/descriptions/timestamp_description.cpp#L36

(I set this value to 6 because Microsoft's ODBC driver complained when it was set to 0 and got fractional seconds for timestamp parameters.)

**Comment by TWAC** [comment]
0 gets another error.
```
Traceback (most recent call last):
  File "C:\python36\lib\site-packages\turbodbc\exceptions.py", line 31, in wrapper
    return f(*args, **kwds)
  File "C:\python36\lib\site-packages\turbodbc\cursor.py", line 71, in execute
    buffer.flush()
turbodbc_intern.Error: ODBC error
state: 22008
native error code: 0
message: [Microsoft][ODBC SQL Server Driver]Datetime field overflow
```

**Comment by MathMagique** [comment]
You can try setting the value to `3` and then making sure that the your test case only uses up to millisecond precision. Apparently, various MSSQL drivers support various precisions for various field types (see http://stackoverflow.com/questions/40918607/cannot-bind-a-sql-type-timestamp-value-using-odbc-with-ms-sql-server-hy104-inv, for example).

**Comment by TWAC** [comment]
Ugh, digits and column size are dynamic per data source in ODBC, somehow retrievable using SQLGetTypeInfo.
You also seem to have to truncate the values so that additional precision is zeros.

In this particular MSSQL setup only 3 digits are used, and hard coding this works, but will probably break other things.
```diff
diff --git a/cpp/turbodbc/Library/src/descriptions/timestamp_description.cpp b/cpp/turbodbc/Library/src/descriptions/timestamp_description.cpp
index 3eaf7f1..6060438 100644
--- a/cpp/turbodbc/Library/src/descriptions/timestamp_description.cpp
+++ b/cpp/turbodbc/Library/src/descriptions/timestamp_description.cpp
@@ -18,7 +18,7 @@ timestamp_description::~timestamp_description() = default;
 
 std::size_t timestamp_description::do_element_size() const
 {
-	return sizeof(SQL_TIMESTAMP_STRUCT);
+	return 20 + 3;
 }
 
 SQLSMALLINT timestamp_description::do_column_c_type() const
@@ -33,7 +33,7 @@ SQLSMALLINT timestamp_description::do_column_sql_type() const
 
 SQLSMALLINT timestamp_description::do_digits() const
 {
-	return 6;
+	return 3;
 }
 
 type_code timestamp_description::do_get_type_code() const
diff --git a/cpp/turbodbc/Library/src/parameter_sets/set_field.cpp b/cpp/turbodbc/Library/src/parameter_sets/set_field.cpp
index 561c057..4590041 100644
--- a/cpp/turbodbc/Library/src/parameter_sets/set_field.cpp
+++ b/cpp/turbodbc/Library/src/parameter_sets/set_field.cpp
@@ -82,7 +82,7 @@ namespace {
 			destination->minute = time.minutes();
 			destination->second = time.seconds();
 			// map posix_time microsecond precision to SQL nanosecond precision
-			destination->fraction = time.fractional_seconds() * 1000;
+			destination->fraction = (time.fractional_seconds() / 1000) * 1000000;
 
 			destination_.indicator = sizeof(SQL_TIMESTAMP_STRUCT);
 		}
diff --git a/cpp/turbodbc_python/Library/src/determine_parameter_type.cpp b/cpp/turbodbc_python/Library/src/determine_parameter_type.cpp
index 6630234..5b9f1d8 100644
--- a/cpp/turbodbc_python/Library/src/determine_parameter_type.cpp
+++ b/cpp/turbodbc_python/Library/src/determine_parameter_type.cpp
@@ -64,7 +64,7 @@ namespace {
 		d->minute = PyDateTime_DATE_GET_MINUTE(ptr);
 		d->second = PyDateTime_DATE_GET_SECOND(ptr);
 		// map microsecond precision to SQL nanosecond precision
-		d->fraction = PyDateTime_DATE_GET_MICROSECOND(ptr) * 1000;
+		d->fraction = (PyDateTime_DATE_GET_MICROSECOND(ptr) / 1000) * 1000000;
 
 		destination.indicator = sizeof(SQL_TIMESTAMP_STRUCT);
 	}
```

```python
curs.execute("select ?", [datetime.datetime.now()])
print(curs.fetchall())
[[datetime.datetime(2017, 3, 16, 17, 24, 38, 357000)]]

curs.execute("select getdate()")
print(curs.fetchall())
[[datetime.datetime(2017, 3, 16, 17, 24, 38, 873000)]]
```

https://docs.microsoft.com/en-us/sql/odbc/reference/appendixes/decimal-digits
https://docs.microsoft.com/en-us/sql/odbc/reference/appendixes/column-size

**Comment by MathMagique** [comment]
May I suggest to move the timestamp problem to another issue, as it does not seem to have anything to do with Windows? This would retain some focus on getting initial Windows support, and probably non-Windows users might be affected by the timestamp problem as well.

**Comment by TWAC** [comment]
The biggest hurdle left for basic windows support is that libturbodbc uses `ext_modules` for its build. On windows build_ext assumes that it is an actual python module, and tries to export a PyInit. And even if that succeeds, I am not sure how linking to a .pyd would work.

An alternative might be to use `libraries` for libturbodbc, which uses build_clib. Unfortunately that only seems to build static libs.

Right now I just include the libturbodbc sources in turbodbc_intern and turbodbc_numpy_support.

**Comment by MathMagique** [comment]
Maybe `ext_modules` is not the problem here. If I understand [this setuptools code](https://github.com/pypa/setuptools/blob/master/setuptools/command/build_ext.py#L115) correctly, we could

    from setuptools.extension import Extension, Library

Then, the turbodbc library could be defined as a `Library` instead of an `Extension`.

I have no idea if this works, and I was not aware of this option until now. If it works, it could have spared me a few loops to jump through.

**Comment by TWAC** [comment]
I can get Library to work with some changes.
However dlls by default do not export any symbols, so I have to add [dllexport](https://pybind11.readthedocs.io/en/master/advanced/misc.html?highlight=dllexport) all over the code to get it to link. That seems prone to breaking, unless Linux also starts using [-fvisibility=hidden](https://gcc.gnu.org/wiki/Visibility).

**Comment by MathMagique** [comment]
Introducing `-fvisibility=hidden` for Linux and OSX builds seems to make sense in that case.

**Comment by TWAC** [comment]
Scratch that, Library creates static libraries (.a) on Linux, defeating the whole purpose.

**Comment by MathMagique** [comment]
What about passing the `-shared` parameter to `Library` as extra arguments on Linux?

**Comment by TWAC** [comment]
Yes, that will probably work.

Getting the shared library working in both Windows and Linux seems like too much work for initial support. I will just make a special cases for Windows in setup.py instead.

**Comment by MathMagique** [comment]
Experimental Windows support has been released thanks to @TWAC. The initial support is not automatically tested yet and still requires infrastructure for compiling the source distribution.

**Comment by MathMagique** [comment]
I have opened #68 to push forward first-class Windows support.
